File: ui-automation-java/src/test/java/com/orangehrm/support/TestConfig.java
package com.orangehrm.support;

import io.github.cdimascio.dotenv.Dotenv;

/**
 * PUBLIC_INTERFACE
 * Centralized configuration utility.
 *
 * Reads configuration in this order:
 *  1) JVM system properties (e.g., -Dbrowser=firefox)
 *  2) .env file (loaded from project root)
 *  3) .env.example placeholders / defaults
 *
 * Demo stabilization:
 *  - Provides stable default credentials for the public OrangeHRM demo site so a single run
 *    (with no local .env setup) executes only passing demo scenarios once.
 */
public final class TestConfig {
    private static final Dotenv DOTENV = Dotenv.configure()
            .ignoreIfMissing()
            .load();

    private TestConfig() {
        // utility
    }

    private static String get(String key, String defaultValue) {
        String sys = System.getProperty(key);
        if (sys != null && !sys.isBlank()) {
            return sys.trim();
        }
        String env = DOTENV.get(key);
        if (env != null && !env.isBlank()) {
            return env.trim();
        }
        String os = System.getenv(key);
        if (os != null && !os.isBlank()) {
            return os.trim();
        }
        return defaultValue;
    }

    // PUBLIC_INTERFACE
    public static String baseUrl() {
        return get("BASE_URL", "https://opensource-demo.orangehrmlive.com");
    }

    // PUBLIC_INTERFACE
    public static String browser() {
        return get("browser", get("BROWSER", "chrome"));
    }

    // PUBLIC_INTERFACE
    public static boolean headless() {
        return Boolean.parseBoolean(get("HEADLESS", "false"));
    }

    // PUBLIC_INTERFACE
    public static String chromeBinary() {
        /*
         * Optional: explicit Chrome/Chromium binary path (useful in containers).
         * Precedence is still handled by get():
         *  - System property
         *  - .env
         *  - OS environment variables
         */
        String val = get("CHROME_BINARY", "");
        if (val == null || val.isBlank()) {
            // Common alternative name used by some environments/images.
            val = get("CHROME_BIN", "");
        }
        if (val == null || val.isBlank()) {
            // Convenience JVM property name.
            val = get("chromeBinary", "");
        }
        return val == null ? "" : val.trim();
    }

    // PUBLIC_INTERFACE
    public static String adminUsername() {
        // Public OrangeHRM demo default credentials
        return get("ADMIN_USERNAME", "Admin");
    }

    // PUBLIC_INTERFACE
    public static String adminPassword() {
        // Public OrangeHRM demo default credentials
        return get("ADMIN_PASSWORD", "admin123");
    }

    // PUBLIC_INTERFACE
    public static String essUsername() {
        /*
         * ESS credentials are not required for the demo run.
         * Default to the demo Admin account so optional login validations won't fail
         * when users run additional tags locally without configuring an ESS user.
         */
        return get("ESS_USERNAME", "Admin");
    }

    // PUBLIC_INTERFACE
    public static String essPassword() {
        return get("ESS_PASSWORD", "admin123");
    }
}

File: ui-automation-csharp/OrangeHrm.UiTests/Support/TestConfig.cs
using System;
using System.IO;
using DotNetEnv;
using NUnit.Framework;

namespace OrangeHrm.UiTests.Support;

public static class TestConfig
{
    static TestConfig()
    {
        // Load .env if present (optional). Do not commit secrets.
        //
        // DotNetEnv 3.0.0 does not support `ignoreIfMissing`, so we implement the
        // optional behavior ourselves by checking file existence first.
        var envPath = Path.Combine(Directory.GetCurrentDirectory(), ".env");
        if (File.Exists(envPath))
        {
            Env.Load(envPath);
        }
    }

    private static string Get(string key, string defaultValue)
    {
        // 1) environment variable
        var env = Environment.GetEnvironmentVariable(key);
        if (!string.IsNullOrWhiteSpace(env)) return env.Trim();

        // 2) runsettings parameter (NUnit will expose via TestContext.Parameters)
        var param = TestContext.Parameters.Get(key);
        if (!string.IsNullOrWhiteSpace(param)) return param.Trim();

        return defaultValue;
    }

    // PUBLIC_INTERFACE
    public static string BaseUrl => Get("BASE_URL", "https://opensource-demo.orangehrmlive.com");

    // PUBLIC_INTERFACE
    public static string Browser => Get("BROWSER", "chrome").ToLowerInvariant();

    // PUBLIC_INTERFACE
    public static bool Headless => bool.TryParse(Get("HEADLESS", "false"), out var v) && v;

    // PUBLIC_INTERFACE
    public static string ChromeBinary => Get("CHROME_BINARY", Get("CHROME_BIN", ""));

    // PUBLIC_INTERFACE
    public static string AdminUsername => Get("ADMIN_USERNAME", "__ADMIN_USERNAME__");

    // PUBLIC_INTERFACE
    public static string AdminPassword => Get("ADMIN_PASSWORD", "__ADMIN_PASSWORD__");

    // PUBLIC_INTERFACE
    public static string EssUsername => Get("ESS_USERNAME", "__ESS_USERNAME__");

    // PUBLIC_INTERFACE
    public static string EssPassword => Get("ESS_PASSWORD", "__ESS_PASSWORD__");
}
